posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py

```diff
@@ -0,0 +1,342 @@
+                    query_id=query_id,
+                    table_name=TEMPORARY_PDI2_TABLE_NAME,
+                    final=True,
+                    deduplicate=True,
+                )
+            ),
+            AsyncMigrationOperation(
+                fn=lambda query_id: run_optimize_table(
+                    unique_name="0006_persons_and_groups_on_events_backfill_groups",
+                    query_id=query_id,
+                    table_name=TEMPORARY_GROUPS_TABLE_NAME,
+                    final=True,
+                    deduplicate=True,
+                    per_shard=True,
+                )
+            ),
+            AsyncMigrationOperationSQL(
+                sql=f"""
+                    ALTER TABLE {TEMPORARY_PDI2_TABLE_NAME} {{on_cluster_clause}}
+                    DELETE WHERE is_deleted = 1 OR person_id IN (
+                        SELECT id FROM {TEMPORARY_PERSONS_TABLE_NAME} WHERE is_deleted=1
+                    )
+                """,
+                sql_settings={"mutations_sync": 2},
+                rollback=None,
+                per_shard=True,
+            ),
+            AsyncMigrationOperationSQL(
+                sql=f"""
+                    ALTER TABLE {TEMPORARY_PERSONS_TABLE_NAME} {{on_cluster_clause}}
+                    DELETE WHERE is_deleted = 1
+                """,
+                sql_settings={"mutations_sync": 2},
+                rollback=None,
+                per_shard=True,
+            ),
+            AsyncMigrationOperationSQL(
+                sql=f"""
+                    CREATE DICTIONARY IF NOT EXISTS person_dict {{on_cluster_clause}}
+                    (
+                        team_id Int64,
+                        id UUID,
+                        properties String,
+                        created_at DateTime
+                    )
+                    PRIMARY KEY team_id, id
+                    SOURCE(CLICKHOUSE(TABLE {TEMPORARY_PERSONS_TABLE_NAME} {self._dictionary_connection_string()}))
+                    LAYOUT(complex_key_cache(size_in_cells 5000000 max_threads_for_updates 6 allow_read_expired_keys 1))
+                    Lifetime(60000)
+                """,
+                rollback=f"DROP DICTIONARY IF EXISTS person_dict {{on_cluster_clause}}",
+                per_shard=True,
+            ),
+            AsyncMigrationOperationSQL(
+                sql=f"""
+                    CREATE DICTIONARY IF NOT EXISTS person_distinct_id2_dict {{on_cluster_clause}}
+                    (
+                        team_id Int64,
+                        distinct_id String,
+                        person_id UUID
+                    )
+                    PRIMARY KEY team_id, distinct_id
+                    SOURCE(CLICKHOUSE(TABLE {TEMPORARY_PDI2_TABLE_NAME} {self._dictionary_connection_string()}))
+                    LAYOUT(complex_key_cache(size_in_cells 50000000 max_threads_for_updates 6 allow_read_expired_keys 1))
+                    Lifetime(60000)
```

Comment:
q: just curious, did you ever experiment with different lifetimes?

Reply:
Nope. It didn't matter for the tests I ran w/ 200M events. What are your thoughts on this, how could we make use of this?